src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -0,0 +1,309 @@
+        /// however is NOT passed on to the children of this activity.  (see Baggage)
+        /// </summary>
+        public IEnumerable<KeyValuePair<string, string>> Tags
+        {
+            get
+            {
+                if (ParentId != null)
+                    yield return new KeyValuePair<string, string>("ParentId", ParentId);
+                for (var tags = _tags; tags != null; tags = tags.Next)
+                    yield return tags.keyValue;
+            }
+        }
+
+        /// <summary>
+        /// Tags are string-string key-value pairs that represent information that will
+        /// be passed along to children of this activity.   Baggage is serialized 
+        /// when requests leave the process (along with the ID).   Typically Baggage is
+        /// used to do fine-grained control over logging of the activty and any children.  
+        /// In general, if you are not using the data at runtime, you should be using Tags 
+        /// instead. 
+        /// </summary> 
+        public IEnumerable<KeyValuePair<string, string>> Baggage
+        {
+            get
+            {
+                for (var activity = this; activity != null; activity = activity.Parent)
+                    for (var baggage = activity._baggage; baggage != null; baggage = baggage.Next)
+                        yield return baggage.keyValue;
+            }
+        }
+
+        /// <summary>
+        /// Returns the value of the key-value pair added to the activity with 'WithBaggage'.
+        /// Returns null if that key does not exist.  
+        /// </summary>
+        public string GetBaggageItem(string key)
+        {
+            foreach (var keyValue in Baggage)
+                if (key == keyValue.Key)
+                    return keyValue.Value;
+            return null;
+        }
+
+        /* Constructors  Builder methods */
+
+        /// <summary>
+        /// Note that Activity has a 'builder' pattern, where you call the constructor, a number of 'With*' APIs and then
+        /// call 'Activity.Start' to build the activity.   You MUST call Start before using it 
+        /// </summary>
+        /// <param name="operationName">Operations name <see cref="OperationName"/></param>
+        public Activity(string operationName)
+        {
+            OperationName = operationName;
+        }
+
+        /// <summary>
+        /// Update the Activity to have a tag with an additional 'key' and value 'value'.
+        /// This shows up in the 'Tags' eumeration.   It is meant for information that
+        /// is useful to log but not needed for runtime control (for the latter, use Baggage)
+        /// </summary>
+        /// <returns>'this' for convinient chaining</returns>
+        public Activity AddTag(string key, string value)
+        {
+            _tags = new KeyValueListNode() { keyValue = new KeyValuePair<string, string>(key, value), Next = _tags };
```

Comment:
Note that the implementation of baggage as a linked list does allow overwriting keys (after all you can call 'AddBaggage' at any time and it will add the new value to the list.   

We chose this data structure because it was efficient it the most likely scenarios (a very small number of unique entries which can be passed very efficiently (no copy, simply reference), to any sub-activities).     

It is true, however that the current implementation leaves the old value in the list.  This does not  affect the GetBaggageItem() (which gets the most recently added key), (which is the primary use for baggage, control at runtime), but does affect the 'Baggage' IEnumerable.     This can of course be fixed, but having but does complicate the implementation.   It is worth consideration...